Installation error latest version of node-red-contrib-meshtastic 2.2.15-3

A little unsure as to where the correct place to post this is, have also posted this error at https://github.com/scruplelesswizard/meshtastic-node-red/issues
Please feel free do delete if this is the wrong place.

Currently trying to install the updated  meshtastic-node-red via the palette or via pnpm throws the following error:

 ` ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@buf%2Fmeshtastic_protobufs.bufbuild_es: Not Found - 404
This error happened while installing the dependencies of @meshtastic/meshtasticjs@2.2.15-1
@buf/meshtastic_protobufs.bufbuild_es is not in the npm registry, or you have no permission to fetch it.
No authorization header was set for the request.
Progress: resolved 4, reused 0, downloaded 3, added 0`

'@buf/meshtastic_protobufs.bufbuild_es@1.5.1-20231206215336-8002ce9b3c7a.1' is not in this registry.

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -20,18 +20,3 @@ jobs:
       - name: Build Package
         run: pnpm build
 
-      - name: Check if version has been updated
-        id: check-version-changed
-        uses: EndBug/version-check@v2
-        with:
-          diff-search: true
-
-      - name: Log when changed
-        if: steps.check-version-changed.outputs.changed == 'true'
-        run: echo "Version change found in commit ${{ steps.check-version-changed.outputs.commit }}! New version is ${{ steps.check-version-changed.outputs.version }} (${{ steps.check-version-changed.outputs.type }})"
-      
-      - name: Log when unchanged
-        if: steps.check-version-changed.outputs.changed == 'false'
-        run: |
-          echo "No version change. Please update the package version in package.json"
-          exit 1

--- a/README.md
+++ b/README.md
@@ -18,3 +18,43 @@ A Node-RED node for decoding Meshtastic Protobufs
 pnpm i
 pnpm build
 ```
+
+## Updating from version 1.x.x to version 2.2.15 or newer
+
+The data structures have changed in versions 2.2.15 from the data in versions 1.x.x. 
+
+Message payloads from version 2 have the following data structure, where `decoded` is the transmitted message:
+
+```
+{
+  "packet": {
+    "from": 1234567890,
+    "to": 9876543210,
+    "channel": 0,
+    "decoded": {
+      "portnum": 5,
+      "payload": {
+        "errorReason": 0
+      },
+      "wantResponse": false,
+      "dest": 0,
+      "source": 0,
+      "requestId": 2345678,
+      "replyId": 0,
+      "emoji": 0
+    },
+    "id": 56789012,
+    "rxTime": 45678901,
+    "rxSnr": 0,
+    "hopLimit": 3,
+    "wantAck": false,
+    "priority": 120,
+    "rxRssi": 0,
+    "delayed": 0
+  },
+  "channelId": "LongFast",
+  "gatewayId": "!abcd1234"
+}
+```
+
+Note that `packet.payloadVariant` is no longer available. Decoded payloads are now available in `packet.decoded`, which retains the same data as `payloadVariant` in a different structure

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@meshtastic/node-red-contrib-meshtastic",
-  "version": "2.2.15-3",
+  "version": "2.2.15-4",
   "description": "Meshtastic protobuf converter for NodeRED",
   "keywords": ["node-red"],
   "scripts": {

--- a/package.json
+++ b/package.json
@@ -2,7 +2,9 @@
   "name": "@meshtastic/node-red-contrib-meshtastic",
   "version": "2.2.15-4",
   "description": "Meshtastic protobuf converter for NodeRED",
-  "keywords": ["node-red"],
+  "keywords": [
+    "node-red"
+  ],
   "scripts": {
     "build": "vite build"
   },
@@ -14,8 +16,12 @@
       "decode": "./dist/decode.js"
     }
   },
-  "files": ["dist", ".npmrc"],
+  "files": [
+    "dist",
+    ".npmrc"
+  ],
   "dependencies": {
+    "@bufbuild/protobuf": "^1.5.1",
     "@meshtastic/meshtasticjs": "2.2.15-1"
   },
   "devDependencies": {

--- a/src/decode.ts
+++ b/src/decode.ts
@@ -13,7 +13,7 @@ const nodeInit: NodeInitializer = (red): void => {
           return;
         }
 
-        var out;
+        let out: any;
 
         switch (decoded.packet.payloadVariant.case) {
           case "encrypted":
@@ -21,7 +21,7 @@ const nodeInit: NodeInitializer = (red): void => {
 
           case "decoded": {
             const { payload } = decoded.packet.payloadVariant.value;
-            const JsonWriteOptions = { "emitDefaultValues": true, "enumAsInteger": true };
+            const jsonWriteOptions = { emitDefaultValues: true, enumAsInteger: true };
             let data: unknown;
 
             switch (decoded.packet.payloadVariant.value.portnum) {
@@ -35,27 +35,27 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.REMOTE_HARDWARE_APP: {
-                data = Protobuf.HardwareMessage.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.HardwareMessage.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.POSITION_APP: {
-                data = Protobuf.Position.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.Position.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.NODEINFO_APP: {
-                data = Protobuf.User.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.User.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.ROUTING_APP: {
-                data = Protobuf.Routing.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.Routing.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.ADMIN_APP: {
-                data = Protobuf.AdminMessage.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.AdminMessage.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
@@ -68,7 +68,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.WAYPOINT_APP: {
-                data = Protobuf.Waypoint.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.Waypoint.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
@@ -94,7 +94,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.STORE_FORWARD_APP: {
-                data = Protobuf.StoreAndForward.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.StoreAndForward.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
@@ -104,7 +104,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.TELEMETRY_APP: {
-                data = Protobuf.Telemetry.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.Telemetry.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
@@ -121,7 +121,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.NEIGHBORINFO_APP: {
-                data = Protobuf.NeighborInfo.fromBinary(payload).toJson(JsonWriteOptions);
+                data = Protobuf.NeighborInfo.fromBinary(payload).toJson(jsonWriteOptions);
                 break;
               }
 
@@ -136,16 +136,16 @@ const nodeInit: NodeInitializer = (red): void => {
 
             console.log(data);
 
-            out = decoded.toJson(JsonWriteOptions)
+            out = decoded.toJson(jsonWriteOptions)
 
             if (data) {
               out.packet.decoded.payload = data;
             }
 
             // include packet.payloadVariant for legacy flows
-            var payloadVariant = {
-              "oneofKind": "decoded",
-              "decoded": out.packet.decoded
+            const payloadVariant = {
+              oneofKind: "decoded",
+              decoded: out.packet.decoded
             };
             out.packet.payloadVariant = payloadVariant;
 

--- a/src/decode.ts
+++ b/src/decode.ts
@@ -1,5 +1,7 @@
 import { Node, NodeDef, NodeInitializer } from "node-red";
 
+import { JsonValue, Message } from "@bufbuild/protobuf"
+
 import { Protobuf } from "@meshtastic/meshtasticjs";
 
 const nodeInit: NodeInitializer = (red): void => {
@@ -13,7 +15,7 @@ const nodeInit: NodeInitializer = (red): void => {
           return;
         }
 
-        let out: any;
+        let out: JsonValue;
 
         switch (decoded.packet.payloadVariant.case) {
           case "encrypted":
@@ -22,7 +24,7 @@ const nodeInit: NodeInitializer = (red): void => {
           case "decoded": {
             const { payload } = decoded.packet.payloadVariant.value;
             const jsonWriteOptions = { emitDefaultValues: true, enumAsInteger: true };
-            let data: unknown;
+            let data: Message | string;
 
             switch (decoded.packet.payloadVariant.value.portnum) {
               case Protobuf.PortNum.UNKNOWN_APP: {
@@ -35,27 +37,27 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.REMOTE_HARDWARE_APP: {
-                data = Protobuf.HardwareMessage.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.HardwareMessage;
                 break;
               }
 
               case Protobuf.PortNum.POSITION_APP: {
-                data = Protobuf.Position.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.Position;
                 break;
               }
 
               case Protobuf.PortNum.NODEINFO_APP: {
-                data = Protobuf.User.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.User;
                 break;
               }
 
               case Protobuf.PortNum.ROUTING_APP: {
-                data = Protobuf.Routing.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.Routing;
                 break;
               }
 
               case Protobuf.PortNum.ADMIN_APP: {
-                data = Protobuf.AdminMessage.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.AdminMessage;
                 break;
               }
 
@@ -68,7 +70,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.WAYPOINT_APP: {
-                data = Protobuf.Waypoint.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.Waypoint;
                 break;
               }
 
@@ -94,7 +96,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.STORE_FORWARD_APP: {
-                data = Protobuf.StoreAndForward.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.StoreAndForward;
                 break;
               }
 
@@ -104,7 +106,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.TELEMETRY_APP: {
-                data = Protobuf.Telemetry.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.Telemetry;
                 break;
               }
 
@@ -121,7 +123,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.NEIGHBORINFO_APP: {
-                data = Protobuf.NeighborInfo.fromBinary(payload).toJson(jsonWriteOptions);
+                data = new Protobuf.NeighborInfo;
                 break;
               }
 
@@ -134,13 +136,24 @@ const nodeInit: NodeInitializer = (red): void => {
               }
             }
 
-            console.log(data);
-
             out = decoded.toJson(jsonWriteOptions)
 
-            if (data) {
-              out.packet.decoded.payload = data;
+            if (!data) {
+              break;
             }
+
+            let decodedPayload: JsonValue | string;
+
+            if (data instanceof Message) {
+              decodedPayload = data.fromBinary(payload).toJson(jsonWriteOptions)
+            } else {
+              decodedPayload = data
+            }
+
+            console.log(decodedPayload);
+
+
+            out.packet.decoded.payload = decodedPayload;
 
             // include packet.payloadVariant for legacy flows
             const payloadVariant = {
@@ -148,7 +161,6 @@ const nodeInit: NodeInitializer = (red): void => {
               decoded: out.packet.decoded
             };
             out.packet.payloadVariant = payloadVariant;
-
             break;
           }
         }

--- a/src/decode.ts
+++ b/src/decode.ts
@@ -14,9 +14,6 @@ const nodeInit: NodeInitializer = (red): void => {
         if (!decoded.packet) {
           return;
         }
-
-        let out: JsonValue;
-
         switch (decoded.packet.payloadVariant.case) {
           case "encrypted":
             break;
@@ -100,7 +97,7 @@ const nodeInit: NodeInitializer = (red): void => {
                 break;
               }
 
-              case Protobuf.PortNum.RANGE_TEST_APP: {
+              case Protobuf.PortNum.dataANGE_TEST_APP: {
                 data = new TextDecoder("ascii").decode(payload);
                 break;
               }
@@ -136,7 +133,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
             }
 
-            out = decoded.toJson(jsonWriteOptions)
+            const out = decoded.toJson(jsonWriteOptions)
 
             if (!data) {
               break;
@@ -152,15 +149,8 @@ const nodeInit: NodeInitializer = (red): void => {
 
             console.log(decodedPayload);
 
-
             out.packet.decoded.payload = decodedPayload;
 
-            // include packet.payloadVariant for legacy flows
-            const payloadVariant = {
-              oneofKind: "decoded",
-              decoded: out.packet.decoded
-            };
-            out.packet.payloadVariant = payloadVariant;
             break;
           }
         }

--- a/src/decode.ts
+++ b/src/decode.ts
@@ -1,6 +1,32 @@
 import { Node, NodeDef, NodeInitializer } from "node-red";
-import { Message } from "@bufbuild/protobuf"
+import { Message } from "@bufbuild/protobuf";
 import { Protobuf } from "@meshtastic/meshtasticjs";
+
+const decoders = {
+  [Protobuf.PortNum.UNKNOWN_APP]: null,
+  [Protobuf.PortNum.TEXT_MESSAGE_APP]: new TextDecoder(),
+  [Protobuf.PortNum.REMOTE_HARDWARE_APP]: new Protobuf.HardwareMessage(),
+  [Protobuf.PortNum.POSITION_APP]: new Protobuf.Position(),
+  [Protobuf.PortNum.NODEINFO_APP]: new Protobuf.User(),
+  [Protobuf.PortNum.ROUTING_APP]: new Protobuf.Routing(),
+  [Protobuf.PortNum.ADMIN_APP]: new Protobuf.AdminMessage(),
+  [Protobuf.PortNum.TEXT_MESSAGE_COMPRESSED_APP]: null,
+  [Protobuf.PortNum.WAYPOINT_APP]: new Protobuf.Waypoint(),
+  [Protobuf.PortNum.AUDIO_APP]: null,
+  [Protobuf.PortNum.DETECTION_SENSOR_APP]: new TextDecoder(),
+  [Protobuf.PortNum.REPLY_APP]: new TextDecoder("ascii"),
+  [Protobuf.PortNum.IP_TUNNEL_APP]: null,
+  [Protobuf.PortNum.SERIAL_APP]: null,
+  [Protobuf.PortNum.STORE_FORWARD_APP]: new Protobuf.StoreAndForward(),
+  [Protobuf.PortNum.RANGE_TEST_APP]: new TextDecoder("ascii"),
+  [Protobuf.PortNum.TELEMETRY_APP]: new Protobuf.Telemetry(),
+  [Protobuf.PortNum.ZPS_APP]: null,
+  [Protobuf.PortNum.SIMULATOR_APP]: null,
+  [Protobuf.PortNum.TRACEROUTE_APP]: null,
+  [Protobuf.PortNum.NEIGHBORINFO_APP]: new Protobuf.NeighborInfo(),
+  [Protobuf.PortNum.PRIVATE_APP]: null,
+  [Protobuf.PortNum.ATAK_FORWARDER]: null,
+};
 
 const nodeInit: NodeInitializer = (red): void => {
   function DecodeNodeConstructor(this: Node, config: NodeDef): void {
@@ -13,7 +39,10 @@ const nodeInit: NodeInitializer = (red): void => {
           return;
         }
 
-        const jsonWriteOptions = { emitDefaultValues: true, enumAsInteger: true };
+        const jsonWriteOptions = {
+          emitDefaultValues: true,
+          enumAsInteger: true,
+        };
         const out = decoded.toJson(jsonWriteOptions);
 
         switch (decoded.packet.payloadVariant.case) {
@@ -21,130 +50,27 @@ const nodeInit: NodeInitializer = (red): void => {
             break;
 
           case "decoded": {
-            const { payload } = decoded.packet.payloadVariant.value;
-            let data: Message | string;
-
-            switch (decoded.packet.payloadVariant.value.portnum) {
-              case Protobuf.PortNum.UNKNOWN_APP: {
+            try {
+              const portNum = decoded.packet.payloadVariant.value.portnum;
+              if (decoders[portNum] === null) {
                 break;
               }
 
-              case Protobuf.PortNum.TEXT_MESSAGE_APP: {
-                data = new TextDecoder().decode(payload);
-                break;
+              const { payload } = decoded.packet.payloadVariant.value;
+              const decoder = decoders[portNum];
+
+              if (decoder instanceof Message) {
+                out.packet.decoded.payload = decoder
+                  .fromBinary(payload)
+                  .toJson(jsonWriteOptions);
+              } else if (decoder instanceof TextDecoder) {
+                out.packet.decoded.payload = decoder.decode(payload);
               }
 
-              case Protobuf.PortNum.REMOTE_HARDWARE_APP: {
-                data = new Protobuf.HardwareMessage;
-                break;
-              }
-
-              case Protobuf.PortNum.POSITION_APP: {
-                data = new Protobuf.Position;
-                break;
-              }
-
-              case Protobuf.PortNum.NODEINFO_APP: {
-                data = new Protobuf.User;
-                break;
-              }
-
-              case Protobuf.PortNum.ROUTING_APP: {
-                data = new Protobuf.Routing;
-                break;
-              }
-
-              case Protobuf.PortNum.ADMIN_APP: {
-                data = new Protobuf.AdminMessage;
-                break;
-              }
-
-              case Protobuf.PortNum.TEXT_MESSAGE_COMPRESSED_APP: {
-                // should never get here
-                // should be decompressed by the firmware to a TEXT_MESSAGE_APP packet
-                throw new Error(
-                  "Received a TEXT_MESSAGE_COMPRESSED_APP message.\nPlease open an issue on Github as this should never happen",
-                );
-              }
-
-              case Protobuf.PortNum.WAYPOINT_APP: {
-                data = new Protobuf.Waypoint;
-                break;
-              }
-
-              case Protobuf.PortNum.AUDIO_APP: {
-                break;
-              }
-
-              case Protobuf.PortNum.DETECTION_SENSOR_APP: {
-                data = new TextDecoder().decode(payload);
-                break;
-              }
-
-              case Protobuf.PortNum.REPLY_APP: {
-                data = new TextDecoder("ascii").decode(payload);
-                break;
-              }
-              case Protobuf.PortNum.IP_TUNNEL_APP: {
-                break;
-              }
-
-              case Protobuf.PortNum.SERIAL_APP: {
-                break;
-              }
-
-              case Protobuf.PortNum.STORE_FORWARD_APP: {
-                data = new Protobuf.StoreAndForward;
-                break;
-              }
-
-              case Protobuf.PortNum.RANGE_TEST_APP: {
-                data = new TextDecoder("ascii").decode(payload);
-                break;
-              }
-
-              case Protobuf.PortNum.TELEMETRY_APP: {
-                data = new Protobuf.Telemetry;
-                break;
-              }
-
-              case Protobuf.PortNum.ZPS_APP: {
-                break;
-              }
-
-              case Protobuf.PortNum.SIMULATOR_APP: {
-                break;
-              }
-
-              case Protobuf.PortNum.TRACEROUTE_APP: {
-                break;
-              }
-
-              case Protobuf.PortNum.NEIGHBORINFO_APP: {
-                data = new Protobuf.NeighborInfo;
-                break;
-              }
-
-              case Protobuf.PortNum.PRIVATE_APP: {
-                break;
-              }
-
-              case Protobuf.PortNum.ATAK_FORWARDER: {
-                break;
-              }
+              console.log(JSON.stringify(out, null, 2));
+            } catch (error) {
+              console.log(`could not decode payload:${error}`);
             }
-
-            if (!data) {
-              break;
-            }
-
-            if (data instanceof Message) {
-              out.packet.decoded.payload = data.fromBinary(payload).toJson(jsonWriteOptions);
-            } else {
-              out.packet.decoded.payload = data;
-            }
-
-            console.log(JSON.stringify(out, null, 2));
 
             break;
           }

--- a/src/decode.ts
+++ b/src/decode.ts
@@ -12,12 +12,16 @@ const nodeInit: NodeInitializer = (red): void => {
         if (!decoded.packet) {
           return;
         }
+
+        var out;
+
         switch (decoded.packet.payloadVariant.case) {
           case "encrypted":
             break;
 
           case "decoded": {
             const { payload } = decoded.packet.payloadVariant.value;
+            const JsonWriteOptions = { "emitDefaultValues": true, "enumAsInteger": true };
             let data: unknown;
 
             switch (decoded.packet.payloadVariant.value.portnum) {
@@ -31,27 +35,27 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.REMOTE_HARDWARE_APP: {
-                data = Protobuf.HardwareMessage.fromBinary(payload);
+                data = Protobuf.HardwareMessage.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.POSITION_APP: {
-                data = Protobuf.Position.fromBinary(payload);
+                data = Protobuf.Position.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.NODEINFO_APP: {
-                data = Protobuf.User.fromBinary(payload);
+                data = Protobuf.User.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.ROUTING_APP: {
-                data = Protobuf.Routing.fromBinary(payload);
+                data = Protobuf.Routing.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
               case Protobuf.PortNum.ADMIN_APP: {
-                data = Protobuf.AdminMessage.fromBinary(payload);
+                data = Protobuf.AdminMessage.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
@@ -64,7 +68,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.WAYPOINT_APP: {
-                data = Protobuf.Waypoint.fromBinary(payload);
+                data = Protobuf.Waypoint.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
@@ -90,7 +94,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.STORE_FORWARD_APP: {
-                data = Protobuf.StoreAndForward.fromBinary(payload);
+                data = Protobuf.StoreAndForward.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
@@ -100,7 +104,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.TELEMETRY_APP: {
-                data = Protobuf.Telemetry.fromBinary(payload);
+                data = Protobuf.Telemetry.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
@@ -117,7 +121,7 @@ const nodeInit: NodeInitializer = (red): void => {
               }
 
               case Protobuf.PortNum.NEIGHBORINFO_APP: {
-                data = Protobuf.NeighborInfo.fromBinary(payload);
+                data = Protobuf.NeighborInfo.fromBinary(payload).toJson(JsonWriteOptions);
                 break;
               }
 
@@ -132,16 +136,25 @@ const nodeInit: NodeInitializer = (red): void => {
 
             console.log(data);
 
+            out = decoded.toJson(JsonWriteOptions)
+
             if (data) {
-              (decoded.packet.payloadVariant.value.payload as unknown) = data;
+              out.packet.decoded.payload = data;
             }
+
+            // include packet.payloadVariant for legacy flows
+            var payloadVariant = {
+              "oneofKind": "decoded",
+              "decoded": out.packet.decoded
+            };
+            out.packet.payloadVariant = payloadVariant;
 
             break;
           }
         }
 
         send({
-          payload: decoded,
+          payload: out,
         });
       }
 

--- a/src/decode.ts
+++ b/src/decode.ts
@@ -1,7 +1,5 @@
 import { Node, NodeDef, NodeInitializer } from "node-red";
-
-import { JsonValue, Message } from "@bufbuild/protobuf"
-
+import { Message } from "@bufbuild/protobuf"
 import { Protobuf } from "@meshtastic/meshtasticjs";
 
 const nodeInit: NodeInitializer = (red): void => {
@@ -14,13 +12,16 @@ const nodeInit: NodeInitializer = (red): void => {
         if (!decoded.packet) {
           return;
         }
+
+        const jsonWriteOptions = { emitDefaultValues: true, enumAsInteger: true };
+        const out = decoded.toJson(jsonWriteOptions);
+
         switch (decoded.packet.payloadVariant.case) {
           case "encrypted":
             break;
 
           case "decoded": {
             const { payload } = decoded.packet.payloadVariant.value;
-            const jsonWriteOptions = { emitDefaultValues: true, enumAsInteger: true };
             let data: Message | string;
 
             switch (decoded.packet.payloadVariant.value.portnum) {
@@ -97,7 +98,7 @@ const nodeInit: NodeInitializer = (red): void => {
                 break;
               }
 
-              case Protobuf.PortNum.dataANGE_TEST_APP: {
+              case Protobuf.PortNum.RANGE_TEST_APP: {
                 data = new TextDecoder("ascii").decode(payload);
                 break;
               }
@@ -133,23 +134,17 @@ const nodeInit: NodeInitializer = (red): void => {
               }
             }
 
-            const out = decoded.toJson(jsonWriteOptions)
-
             if (!data) {
               break;
             }
 
-            let decodedPayload: JsonValue | string;
-
             if (data instanceof Message) {
-              decodedPayload = data.fromBinary(payload).toJson(jsonWriteOptions)
+              out.packet.decoded.payload = data.fromBinary(payload).toJson(jsonWriteOptions);
             } else {
-              decodedPayload = data
+              out.packet.decoded.payload = data;
             }
 
-            console.log(decodedPayload);
-
-            out.packet.decoded.payload = decodedPayload;
+            console.log(JSON.stringify(out, null, 2));
 
             break;
           }

